src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1395,6 +1405,12 @@ MulticopterPositionControl::task_main()
 				if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.valid
 						&& _pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_LAND) {
 					_vel_sp(2) = _params.land_speed;
+
+					/* final approach, decrease descent speed */
+					if(_params.land_final_approach_alt > 0.0f && _local_pos.z_valid
```

Comment:
For using the range finder you can add a condition before this where you check if the terrain estimate is valid and then calculate the altitude difference from there. These are the relevant fields https://github.com/PX4/Firmware/blob/master/msg/vehicle_global_position.msg#L18-L19 (not in this case it's global altitude)